test/mainnet-fork/Uniswapv4.t.sol

```diff
@@ -864,7 +864,109 @@ contract MainnetControllerUniswapV4Tests is ForkTestBase {
+
+        tickUpper = int24(_bound(int256(tickUpper), int256(tickLower) + 1, boundedUpperMax));
+        initialLiquidity = uint128(_bound(uint256(initialLiquidity), 1e6, 2_000_000e6));
+
+        uint128 additionalLiquidity = initialLiquidity / 2;
+
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setUniswapV4TickLimits(_POOL_ID, -10_000, 10_000, 1_000);
+        rateLimits.setRateLimitData(_DEPOSIT_LIMIT_KEY,  2_000_000e18, uint256(2_000_000e18) / 1 days);
+        rateLimits.setRateLimitData(_WITHDRAW_LIMIT_KEY, 2_000_000e18, uint256(2_000_000e18) / 1 days);
+        vm.stopPrank();
+
+        IncreasePositionResult memory mintResult     = _mintPosition(tickLower, tickUpper, initialLiquidity, type(uint160).max, type(uint160).max);
+        IncreasePositionResult memory increaseResult = _increasePosition(mintResult.tokenId, additionalLiquidity, type(uint160).max, type(uint160).max);
+
+        uint256 valueBeforeIncrease = _to18From6Decimals(mintResult.amount0Spent)     + _to18From6Decimals(mintResult.amount1Spent);
+        uint256 valueAdded          = _to18From6Decimals(increaseResult.amount0Spent) + _to18From6Decimals(increaseResult.amount1Spent);
+        uint256 totalValueDeposited = valueBeforeIncrease + valueAdded;
+
+        uint128 totalLiquidity = mintResult.liquidityIncrease + increaseResult.liquidityIncrease;
+
+        DecreasePositionResult memory decreaseResult = _decreasePosition(mintResult.tokenId, totalLiquidity, 0, 0);
+
+        uint256 valueReceived  = _to18From6Decimals(decreaseResult.amount0Received) + _to18From6Decimals(decreaseResult.amount1Received);
+
+        // Value difference should be less than 1e12 which including protocol fees and rounding errors
+        // 10 wei at 6 decimals is 1e13 at 18 decimals
+        assertApproxEqAbs(totalValueDeposited, valueReceived, 1e13);
+    }
+
+    function testFuzz_swapUniswapV4_amounts(
+        uint128 amountIn,
+        bool    swapDirection  // true = USDC->USDT, false = USDT->USDC
+    ) external {
+        amountIn = uint128(_bound(uint256(amountIn), 1e6, 1_000_000e6));
+
+        vm.startPrank(SPARK_PROXY);
+        mainnetController.setMaxSlippage(address(uint160(uint256(_POOL_ID))), 0.98e18);
+        rateLimits.setRateLimitData(_SWAP_LIMIT_KEY, 1_000_000e18, 0);
+        vm.stopPrank();
+
+        address tokenIn = swapDirection ? address(usdc) : address(usdt);
+
+        uint128 amountOutMin = _getSwapAmountOutMin(tokenIn, amountIn, 0.99e18);
+
+        uint256 rateLimitBefore = rateLimits.getCurrentRateLimit(_SWAP_LIMIT_KEY);
+
+        uint256 amountOut = _swap(tokenIn, amountIn, amountOutMin);
+
+        assertEq(rateLimits.getCurrentRateLimit(_SWAP_LIMIT_KEY), rateLimitBefore - _to18From6Decimals(amountIn));
+
+        assertGe(amountOut, amountOutMin);
+
+        uint256 normalizedAmountIn  = _to18From6Decimals(amountIn);
+        uint256 normalizedAmountOut = _to18From6Decimals(uint128(amountOut));
```

Comment:
Don't need to normalize these, they're both 18 decimals

Reply:
done